### DISCUSSION TITLE
onremove and key change

### DISCUSSION THREAD

**Discussion by cavemansspa** [MithrilJS/mithril.js#2725] [Q&A]
Should `onremove` be invoked in this example: https://jsfiddle.net/6dysnpkL/

```
let count = 0
m.mount(document.body, {

  view: (vnode) => m('div', {
    key: count, // key changes, view onupdate result in console output
    onremove: (vnode) => {
      console.log('onremove')  // not invoked
    },
    onupdate: (vnode) => {
      console.log('onupdate', vnode)
    },

  }, 'hello ' + count)
})

let id = setInterval(() => {
  count++
  console.log('counter', count)
  if(count > 9) clearInterval(id)
  m.redraw()
}, 1000)
```

**Comment by pygy** [comment]
It's a common gotcha, and it took me a while to detect it as well...

Keys only work in list context. Wrap the `m('div',...)` call in an array and `onremove` fires.

The tricky thing is that the engine implicitly creates lists in some places but not always.

**Comment by barneycarroll** [comment]
The route docs advocate using `key` on single components: http://mithril.js.org/route.html#key-parameter

**Comment by pygy** [comment]
@barneycarroll `m.render(element, vnode)` normalizes `vnode` to `[vnode]`. So keys work for the root element (and that extends to whatever `RouteResolver.render` returns). That's what I meant by *"the engine implicitly creates lists in some places"*, and why I labeled it as a gotcha, the docs must be improved...

**Comment by barneycarroll** [comment]
@pygy the expected behaviour is intuitive, and describing the reasons for the edge case is awkward and unrewarding. How about we change components to always wrap returns in fragments? This would fix the issue transparently and make vnode structure more predictable.

**Comment by barneycarroll** [comment]
@cavemansspa in the meantime, it's easy enough to ensure all element & component nodes are in lists by wrapping view return values in arrays: https://jsfiddle.net/barney/Lsr7be0y/

**Comment by valtron** [comment]
Here's another instance of this: https://jsfiddle.net/u9rgLyk2/3/

My intention for using `key` was to indicate that the new vnode is different, i.e. if keys are different, don't recycle; instead, unmount the old vnode and create a new one. [This](https://github.com/barneycarroll/mithril.js/commit/c573ceb48f3463a2ca6b51c5fde7d318c97e49f4) fixes it nicely.

Another solution would be a new lifecycle method `onmatch(vnode, old)` (or whatever) called [before `shouldNotUpdate`](https://github.com/MithrilJS/mithril.js/blob/1aaa2ff568491104d5e5816463292fdb01ff7c84/render/render.js#L440) which, if `false` is returned, does `removeNode(old); createNode(parent, ...)`. In this case, I wouldn't use `key`: rather I'd compare the relevant vnode attrs (i.e. `options` in the chart example) inside `onmatch`.

**Comment by cavemansspa** [comment]
@valtron -- did you see the gitter followups https://gitter.im/mithriljs/mithril.js?at=5c3125053cf6c35a8497fd6d and https://gitter.im/mithriljs/mithril.js?at=5c31280e1491b27876c818ff?

**Comment by valtron** [comment]
@cavemansspa I did, that's just a workaround though, required by the fact that mixing keyed/unkeyed nodes isn't supported/defined. Would be better, IMO, to define and support this behaviour, since the [change](https://github.com/barneycarroll/mithril.js/commit/c573ceb48f3463a2ca6b51c5fde7d318c97e49f4) is so simple.
